File: examples/csharp/heartbeat_generator/heartbeat_generator.cs
/*
 * Copyright 2007-2023 United States Government as represented by the
 * Administrator of The National Aeronautics and Space Administration.
 * No copyright is claimed in the United States under Title 17, U.S. Code.
 * All Rights Reserved.
 */


/**
 * @file heartbeat_generator.cs
 *
 * @brief This example program provides a basic demonstration of using
 * the HeartbeatGenerator to publish heartbeat messages onto the GMSEC
 * Bus.
 */

using GMSEC.API5;
using System;
using System.Text;


public class heartbeat_generator
{
	public static int Main(string[] args)
	{
		// Set up connection configuration options, some from the command line,
		// others hard-coded.
		Config config = new Config(args);

		// Initialize log level for output
		InitializeLogLevel(config);

		// Validate messages before being sent to the bus
		config.AddValue("gmsec-msg-content-validate-send", "true");

		// Define the heartbeat publish rate
		ushort publishRate = 2;  // in seconds

		try
		{
			// Create HeartbeatGenerator instance.
			using (HeartbeatGenerator hbgen = new HeartbeatGenerator(config, publishRate, GetStandardFields()))
			{
				// Output information regarding the API version and connection
				Log.Info(Connection.GetAPIVersion());
				Log.Info("HeartbeatGenerator publish rate is: " + publishRate + " seconds");

				// Start the heartbeat generator
				hbgen.Start();
				Log.Info("HeartbeatGenerator is running; use gmsub or other utility to monitor messages");

				Log.Info("Press <enter> to stop the HeartbeatGenerator");
				Console.ReadLine();

				// Stop the HeartbeatGenerator
				hbgen.Stop();
				Log.Info("HeartbeatGenerator has been stopped");
			}
		}
		catch (GmsecException e)
		{
			// If here, the most likely case is the middleware could not be
			// deduced, or a failure occurred when attempting to connect to
			// the GMSEC Bus.
			Log.Error("GmsecException: " + e.ToString());
			return -1;
		}

		return 0;
	}


	private static void InitializeLogLevel(Config config)
	{
		LogLevel level = Log.FromString( config.GetValue("loglevel", "info") );
		Log.SetReportingLevel(level);
	}


	private static FieldList GetStandardFields()
	{
		FieldList standardFields = new FieldList()
		{
			new StringField("MISSION-ID", "MY-MISSION", true),
			new StringField("SAT-ID-PHYSICAL", "MY-SAT-ID", true),
			new StringField("SAT-ID-LOGICAL", "MY-SAT-ID", true),
			new StringField("CONSTELLATION-ID", "MY-CNST", true),
			new StringField("FACILITY", "MY-FACILITY", true),
			new StringField("DOMAIN1", "MY-DOMAIN-1", true),
			new StringField("DOMAIN2", "MY-DOMAIN-2", true),
			new StringField("COMPONENT", "HB-GEN", true)
		};

		return standardFields;
	}
}

File: examples/c/heartbeat_generator/heartbeat_generator.c
/*
 * Copyright 2007-2023 United States Government as represented by the
 * Administrator of The National Aeronautics and Space Administration.
 * No copyright is claimed in the United States under Title 17, U.S. Code.
 * All Rights Reserved.
 */


/**
 * @file heartbeat_generator.c
 *
 * @brief This example program provides a basic demonstration of using
 * the HeartbeatGenerator to publish heartbeat messages onto the GMSEC
 * Bus.
 */

#include "../utility.h"

#include <gmsec5_c.h>

#include <stdio.h>
#include <stdlib.h>


void checkStatus(const char* operation, GMSEC_Status status);

void getStandardFields(GMSEC_Field** standardFields, size_t* numFields);
void freeStandardFields(GMSEC_Field** standardFields, size_t numFields);


int main(int argc, char** argv)
{
	GMSEC_Status status = statusCreate();

	// Set up connection configuration options, some from the command line,
	// others hard-coded.
	GMSEC_Config config = configCreateWithArgs(argc, argv);

	// Initialize log level for output
	initializeLogLevel(config);

	// Validate messages before being sent to the bus
	configAddValue(config, "gmsec-msg-content-validate-send", "true", NULL);

	// Define the heartbeat publish rate
	GMSEC_U32 publishRate = 2;  // in seconds

	// Acquire standard fields to be included with the heartbeat messages
	GMSEC_Field* standardFields = {0};
	size_t       numFields = 0;

	getStandardFields(&standardFields, &numFields);

	// Create HeartbeatGenerator instance.
	GMSEC_HeartbeatGenerator hbgen = heartbeatGeneratorCreateWithFields(config, publishRate, standardFields, numFields, status);
	checkStatus("heartbeatGeneratorCreateWithFields", status);

	// Release resources associated with standard fields
	freeStandardFields(&standardFields, numFields);

	// Output information regarding the API version and connection
	GMSEC_INFO(connectionGetAPIVersion());
	GMSEC_INFO("HeartbeatGenerator publish rate is: %u seconds", publishRate);

	// Start the heartbeat generator
	heartbeatGeneratorStart(hbgen, status);
	checkStatus("heartbeatGeneratorStart", status);

	GMSEC_INFO("HeartbeatGenerator is running; use gmsub or other utility to monitor messages");

	GMSEC_INFO("Press <enter> to stop the HeartbeatGenerator");
	getchar();

	// Stop the HeartbeatGenerator
	heartbeatGeneratorStop(hbgen, status);
	checkStatus("heartbeatGeneratorStop", status);

	GMSEC_INFO("HeartbeatGenerator has been stopped");

	// Cleanup resources
	heartbeatGeneratorDestroy(&hbgen);
	configDestroy(&config);
	statusDestroy(&status);

	return 0;
}


void checkStatus(const char* operation, GMSEC_Status status)
{
	if (statusHasError(status) == GMSEC_TRUE)
	{
		GMSEC_ERROR("%s failed; reason: %s", operation, statusGet(status));
		exit(-1);
	}
}


void getStandardFields(GMSEC_Field** standardFields, size_t* numFields)
{
	*numFields      = 8;
	*standardFields = malloc(*numFields * sizeof(GMSEC_Field));

	(*standardFields)[0] = stringFieldCreate("MISSION-ID", "MY-MISSION", GMSEC_TRUE, NULL);
	(*standardFields)[1] = stringFieldCreate("SAT-ID-PHYSICAL", "MY-SAT-ID", GMSEC_TRUE, NULL);
	(*standardFields)[2] = stringFieldCreate("SAT-ID-LOGICAL", "MY-SAT-ID", GMSEC_TRUE, NULL);
	(*standardFields)[3] = stringFieldCreate("CONSTELLATION-ID", "MY-CNST", GMSEC_TRUE, NULL);
	(*standardFields)[4] = stringFieldCreate("FACILITY", "MY-FACILITY", GMSEC_TRUE, NULL);
	(*standardFields)[5] = stringFieldCreate("DOMAIN1", "MY-DOMAIN-1", GMSEC_TRUE, NULL);
	(*standardFields)[6] = stringFieldCreate("DOMAIN2", "MY-DOMAIN-2", GMSEC_TRUE, NULL);
	(*standardFields)[7] = stringFieldCreate("COMPONENT", "HB-GEN", GMSEC_TRUE, NULL);
}


void freeStandardFields(GMSEC_Field** standardFields, size_t numFields)
{
	size_t i;

	for (i = 0; i < numFields; ++i)
	{
		fieldDestroy(&(*standardFields)[i]);
	}

	free(*standardFields);

	*standardFields = NULL;
}
